### ISSUE TITLE
gmond plugin causes collectd to crash

### ISSUE THREAD

**Issue by brianpkelly** [collectd/collectd#1129]
Hello,

When I have the gmond plugin enabled, collectd crashes shortly after initialization. I have a `DEBUG` statement before `mc_receive_thread_start ();` in `gmond_init` and another in the beginning of `mc_receive_thread_start`. The log file has the first message, but not the second, and there are no messages after it. A few plugins submit data before collectd crashes. The gmond daemon appears to be working correctly and both collectd and gmond seem to be configured correctly.

Thanks,
Brian

**Comment by octo** [comment]
Hi @brianpkelly, thanks for reporting this issue! Unfortunately, debugging this from a description alone is quite hard. Would you be able to generate a [core dump](https://collectd.org/wiki/index.php/Core_file)? The collectd wiki has an explanation of the steps necessary …

Thanks, —octo

**Comment by brianpkelly** [comment]
Thanks for the reply, @octo. Unfortunately, I'm having issues getting a core dump. I'm using Ubuntu 14.04 and compiling from source using gcc. I also tried installing collectd-dbg, but that didn't help. I can get core dumps from other programs.

**Comment by octo** [comment]
@brianpkelly weird. Well, you can try running collectd inside the debugger:

```
$ cd $prefix
$ gdb sbin/collectd
(gdb) run -f
```

This should run *collectd* in the debugger (in foreground mode). If collectd crashes, the debugger will automatically dump you into a shell where you can then enter

```
(gdb) backtrace full
```

HTH, —octo

**Comment by brianpkelly** [comment]
@octo thanks again. I could not get a stack trace, but here is the gdb session: 
```
...
Reading symbols from /opt/collectd/sbin/collectd...done.
(gdb) run -f
Starting program: /opt/collectd/sbin/collectd -f
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/i386-linux-gnu/libthread_db.so.1".
[New Thread 0xb7d9eb40 (LWP 26729)]
[New Thread 0xb759db40 (LWP 26730)]
[New Thread 0xb6d9cb40 (LWP 26731)]
[New Thread 0xb659bb40 (LWP 26732)]
[New Thread 0xb5d9ab40 (LWP 26733)]
[New Thread 0xb5599b40 (LWP 26734)]
[New Thread 0xb4d98b40 (LWP 26735)]
[New Thread 0xb43ffb40 (LWP 26736)]
[New Thread 0xb39ffb40 (LWP 26737)]
[New Thread 0xb2fffb40 (LWP 26738)]
[New Thread 0xb27feb40 (LWP 26739)]
/opt/collectd/sbin/collectd: symbol lookup error: /opt/collectd/lib/collectd/gmond.so: undefined symbol: xdr_Ganglia_msg_formats
[Thread 0xb5599b40 (LWP 26734) exited]
[Thread 0xb2fffb40 (LWP 26738) exited]
[Thread 0xb7d9eb40 (LWP 26729) exited]
[Thread 0xb7dea700 (LWP 26725) exited]
[Thread 0xb759db40 (LWP 26730) exited]
[Thread 0xb659bb40 (LWP 26732) exited]
[Thread 0xb5d9ab40 (LWP 26733) exited]
[Thread 0xb43ffb40 (LWP 26736) exited]
[Thread 0xb6d9cb40 (LWP 26731) exited]
[Thread 0xb4d98b40 (LWP 26735) exited]
[Thread 0xb39ffb40 (LWP 26737) exited]
[Inferior 1 (process 26725) exited with code 0177]
(gdb) backtrace full
No stack.
(gdb) Quit
```
It looks like someone was having similar issues in the mailing list, but I could not find a resolution. Maybe it's just Ubuntu-specific. Maybe this will help:
```
root@ubuntuvm:/# find -name 'libganglia*'
./usr/lib/libganglia.so
./usr/lib/libganglia.a
./usr/lib/libganglia.la
./usr/lib/libganglia-3.6.0.so.0
./usr/lib/libganglia-3.6.0.so.0.0.0
./usr/share/doc/libganglia1
./usr/share/doc/libganglia1-dev
./var/lib/dpkg/info/libganglia1.md5sums
./var/lib/dpkg/info/libganglia1.list
./var/lib/dpkg/info/libganglia1-dev.md5sums
./var/lib/dpkg/info/libganglia1-dev.list
./var/lib/dpkg/info/libganglia1.postrm
./var/lib/dpkg/info/libganglia1.shlibs
./var/lib/dpkg/info/libganglia1.postinst
```
Thanks again.

**Comment by octo** [comment]
Hi @brianpkelly,

it appears that the daemon tires to load the *gmond plugin* but fails because it has unresolved symbols. You can run:
```bash
ldd /opt/collectd/lib/collectd/gmond.so
```
to get a list of libraries it is looking for.

Since collectd can't load the plugin, it can't do anything harmful like crash the daemon. The logs indicate, however, that the daemon shuts down cleanly. Can you provide a copy of your configuration, ideally stripped of all comments and empty lines?

Thanks, —octo

**Comment by brianpkelly** [comment]
Hey, @octo, I tested it out with a bare-bones configuration to see how it would do:

```
LoadPlugin logfile

<Plugin logfile>
	LogLevel debug
	File "/var/log/collectd.log"
	Timestamp false
	PrintSeverity false
</Plugin>

LoadPlugin gmond

<Plugin "gmond">
   MCReceiveFrom "239.2.11.71" "8649"
   <Metric "swap_total">
     Type "swap"
     TypeInstance "total"
     DataSource "value"
   </Metric>
   <Metric "swap_free">
     Type "swap"
     TypeInstance "free"
     DataSource "value"
   </Metric>
</Plugin>
```
Still shuts down. This is interesting though:
```
user@ubuntuvm:/$ ldd /opt/collectd/lib/collectd/gmond.so 
	linux-gate.so.1 =>  (0xb7779000)
	libc.so.6 => /lib/i386-linux-gnu/libc.so.6 (0xb75ae000)
	/lib/ld-linux.so.2 (0xb777a000)
```
Shouldn't there be a dependency on libganglia? For comparison, gmond's dependencies:
```
user@ubuntuvm:/$ ldd usr/sbin/gmond 
	linux-gate.so.1 =>  (0xb779f000)
	libapr-1.so.0 => /usr/lib/i386-linux-gnu/libapr-1.so.0 (0xb7754000)
	libganglia-3.6.0.so.0 => /usr/lib/libganglia-3.6.0.so.0 (0xb773d000)
	libz.so.1 => /lib/i386-linux-gnu/libz.so.1 (0xb7722000)
	libpcre.so.3 => /lib/i386-linux-gnu/libpcre.so.3 (0xb76e4000)
	libconfuse.so.0 => /usr/lib/i386-linux-gnu/libconfuse.so.0 (0xb76d7000)
	libpthread.so.0 => /lib/i386-linux-gnu/libpthread.so.0 (0xb76bb000)
	libc.so.6 => /lib/i386-linux-gnu/libc.so.6 (0xb750d000)
	libuuid.so.1 => /lib/i386-linux-gnu/libuuid.so.1 (0xb7506000)
	libdl.so.2 => /lib/i386-linux-gnu/libdl.so.2 (0xb7501000)
	libexpat.so.1 => /lib/i386-linux-gnu/libexpat.so.1 (0xb74d8000)
```
Thanks,
Brian

**Comment by mfournier** [comment]
It seems the configure script looks for the `ganglia-config` utility, which isn't available when installing libganglia1-dev (on debian/ubuntu). I think the problem is that the configure script doesn't set the `GANGLIA_*FLAGS` variables properly when `ganglia-config` doesn't exist.

**Comment by mfournier** [comment]
This is fixed by 1a0ffe9, and will be part of upcoming 5.4.3 and 5.5.1 bugfix releases. Thanks for reporting the issue @brianpkelly !
